### PR TITLE
fix: improve convertion errors and handle invalid ABI

### PR DESCRIPTION
**Pull request by whichqua**
Problem: During convertions between types from blockifier, starknet and cairolang classes, the errors returned are not consistent. This also led to finding an invalid api bug during convertion.

Solution:
- Create better errors for `ConvertionError` and `RpcClientError` and handle them appropriately.
- Default to `None` when abi is not valid JSON.

Issue Number: N/A

## Type

- [ ] feature
- [x] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: crates/rpc-client/src/pathfinder/client.rs [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
 use serde::Deserialize;
 use serde_json::json;
 use starknet_types_core::felt::Felt;
 
 use crate::pathfinder::proofs::{PathfinderClassProof, PathfinderProof};
 
+#[derive(Debug, thiserror::Error)]
+pub enum ClientError {
+    #[error("Encountered a request error: {0}")]
+    ReqwestError(#[from] reqwest::Error),
```
Comment (shamsasari) [inline crates/rpc-client/src/pathfinder/client.rs:12]: From what I've seen of Rust errors, shouldn't this just be `Reqwest`? The enum is already named "Error".

File: crates/rpc-client/src/pathfinder/client.rs [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
+    #[error("Encountered a custom error: {0}")]
+    Msg(String),
```
Comment (shamsasari) [inline crates/rpc-client/src/pathfinder/client.rs:14]: `Custom` would be a better variant name

File: crates/starknet-os-types/src/sierra_contract_class.rs [hunk 3]
```diff
@@ -39,12 +39,13 @@ impl GenericSierraContractClass {
...
     fn build_cairo_lang_class(&self) -> Result<CairoLangSierraContractClass, ContractClassError> {
-        if let Ok(serialized_class) = self.get_serialized_contract_class() {
-            let contract_class = serde_json::from_slice(serialized_class)?;
-            return Ok(contract_class);
+        match self.get_serialized_contract_class() {
+            Ok(serialized_class) => {
+                let contract_class = serde_json::from_slice(serialized_class)?;
+                Ok(contract_class)
+            }
+            Err(e) => Err(e),
```
Comment (shamsasari) [inline crates/starknet-os-types/src/sierra_contract_class.rs:47]: I think you can improve this by using `Result.and_then`

File: crates/starknet-os-types/src/sierra_contract_class.rs [hunk 4]
```diff
@@ -140,7 +141,8 @@ impl TryFrom<&FlattenedSierraClass> for FlattenedSierraClassWithAbi {
     type Error = serde_json::error::Error;
 
     fn try_from(sierra_class: &FlattenedSierraClass) -> Result<Self, Self::Error> {
-        let abi: Option<cairo_lang_starknet_classes::abi::Contract> = serde_json::from_str(&sierra_class.abi)?;
+        let abi: Option<cairo_lang_starknet_classes::abi::Contract> =
+            serde_json::from_str(&sierra_class.abi).unwrap_or_default();
```
Comment (HermanObst) [inline crates/starknet-os-types/src/sierra_contract_class.rs:145]: @whichqua what default would represent here?

Comment (HermanObst) [reply crates/starknet-os-types/src/sierra_contract_class.rs:145]: @whichqua so this line is the actual handling of the ABI, right?

File: crates/bin/prove_block/src/rpc_utils.rs [hunk 5]
```diff
@@ -124,11 +125,8 @@ pub(crate) async fn get_storage_proofs(
     let mut storage_proofs = HashMap::new();
 
     log::info!("Contracts we're fetching proofs for:");
-    for contract_address in accessed_keys_by_address.keys() {
-        log::info!("    {}", contract_address.to_string());
-    }
-
     for (contract_address, storage_keys) in accessed_keys_by_address {
+        log::info!("    {}", contract_address.to_string());
```
Comment (notlesh) [inline crates/bin/prove_block/src/rpc_utils.rs:129]: I know this was just moved, but I suggest giving some more context in the log message -- I had a hard time finding this log statement last time I wanted to remove it ;)

**Review by notlesh — APPROVED** [review]

**Comment by HermanObst** [comment]
@whichqua please address @shamsasari comments before merging

File: crates/rpc-client/src/pathfinder/client.rs (hunk 1 above)
Comment (shamsasari) [reply crates/rpc-client/src/pathfinder/client.rs:12]: Wow, definitely don't like that! But at least you've made your variants consistent

**Review by HermanObst — APPROVED** [review]
LGTM

**Pull request merged**